File: Python_Ver/utils/simuData.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.ndimage import gaussian_filter1d
import os
from scipy.io import savemat

def simuData():
        
    # Set random seed for reproducibility
    np.random.seed(0)

    # Pre-define storage lists for variable-length data
    rate_input_list = []  # one mPFC neuron
    rate_output_list = []  # three M1 neurons
    segment_list = []
    trialNo_list = []

    mPFCnum = 1
    M1num = 3
    M1order = [1, 2, 3]
    M1num_pre = 2  # only predict the first two

    trialType = 1

    for trialIdx in range(1, 201):  # simulate 200 trials
        startTime = np.random.randint(40, 51)  # start cue time
        responseTime = np.random.randint(50, 61)
        pressTime = np.random.randint(80, 101)

        trialType = 2 if trialType == 1 else 1

        # mPFC firing changes 200 ms ahead of press movement
        mPFC_part = np.concatenate([
            np.full(startTime + 50 + responseTime - 20, 0.05),
            np.full(20 + pressTime, 0.3 + 0.4 * (trialType == 2))
        ])

        # M1 firing patterns
        M1_part = np.vstack([
            np.concatenate([np.full(startTime + 50 + responseTime, 0.1),
                            np.full(pressTime, 0.1 + 0.7 * (trialType == 2))]),
            np.concatenate([np.full(startTime + 50 + responseTime, 0.05),
                            np.full(pressTime, 0.7)]),
            np.full(startTime + 50 + responseTime + pressTime, 0.5)
        ])

        segment_part = np.concatenate([
            np.zeros(startTime), np.ones(50), np.zeros(responseTime),
            np.full(pressTime, 1 + trialType)
        ])

        trialNo_part = np.concatenate([
            np.zeros(startTime), np.full(50, trialIdx), np.zeros(responseTime),
            np.full(pressTime, trialIdx)
        ])

        # Append new data
        rate_input_list.append(mPFC_part)
        rate_output_list.append(M1_part)
        segment_list.append(segment_part)
        trialNo_list.append(trialNo_part)

    # Convert lists to NumPy arrays
    rate_input = np.concatenate(rate_input_list)
    rate_output = np.hstack(rate_output_list)  # Stack horizontally for (3, N) shape
    segment = np.concatenate(segment_list)
    trialNo = np.concatenate(trialNo_list)

    # Apply Gaussian smoothing
    def gaussian_smooth(data, sigma=10):
        return gaussian_filter1d(data, sigma=sigma, axis=-1)

    rate_input = gaussian_smooth(rate_input)
    mPFC = rate_input > np.random.rand(*rate_input.shape)
    rate_output = gaussian_smooth(rate_output)
    M1 = rate_output > np.random.rand(*rate_output.shape)

    modelName = 'decodingModel_simulation'
    his = 16

    # Plotting
    plotIndex = np.arange(1000)
    timeIndex = plotIndex / 100

    fig, axes = plt.subplots(4, 1, figsize=(8, 10))
    fig.patch.set_facecolor('w')

    axes[0].plot(timeIndex, rate_input[plotIndex], 'k', linewidth=1.5)
    axes[0].set_ylabel('mPFC')
    axes[0].set_title("Close the window to see training results.")

    axes[1].plot(timeIndex, rate_output[0, plotIndex], 'k', linewidth=1.5)
    axes[1].set_ylim([0, 0.8])
    axes[1].set_ylabel('M1 1')

    axes[2].plot(timeIndex, rate_output[1, plotIndex], 'k', linewidth=1.5)
    axes[2].set_ylim([0, 0.8])
    axes[2].set_ylabel('M1 2')

    axes[3].plot(timeIndex, segment[plotIndex], 'k', linewidth=1.5)
    axes[3].set_ylabel('movements')

    # Save data to .mat file
    os.makedirs('data', exist_ok=True)  # Ensure the directory exists
    savemat('Python_Ver/data/simulations.mat', {
        'mPFC': mPFC,
        'mPFCnum': mPFCnum,
        'M1': M1.T,
        'M1num': M1num,
        'M1order': M1order,
        'M1num_pre': M1num_pre,
        'segment': segment,
        'trialNo': trialNo,
        'his': his,
        'modelName': modelName,
        'DataIndex': 'Simulations'
    })

    print("Data saved to simulations.mat")

    return fig, axes
    

File: utils/simuData.m
clear; close all;
DataIndex = 'Simulations';
rng('default');

%% pre-define neural probabilities
% 10ms per time bin

% one mPFC neuron
rate_input = zeros(1, 2e4);
mPFCnum = 1;

% three M1 neuron, which only the first two are related to mPFC
rate_output = zeros(3, 2e4);
M1num = 3;
M1order = [1, 2, 3];
M1num_pre = 2; % only predict the first two

segment = zeros(1, 2e4); % movement of the "rat"
trialNo = zeros(1, 2e4); % trial No, from 1~200

%%
cursor = 1;
trialType = 1;
for trialIdx = 1:200 % simulate 200 trials

  startTime = randi([40, 50]); % start cue given after ([40 50])+50 time bins
  responseTime = randi([50, 60]); 
  pressTime = randi([80, 100]); % hold the level

  % trialType 1 means low trial, 2 means high trial 
  if trialType == 1
    trialType = 2;
  elseif trialType == 2
    trialType = 1;
  end
  
  % assume the mPFC firing changes 200 ms ahead of the press movement
  % baseline firing: 0.05, low lever: 0.3, high lever: 0.7
  mPFC_part = [0.05*ones(1, startTime+50+responseTime-20), (0.3+0.4*(trialType==2))*ones(1, 20+pressTime)];

  % M1 1: 0.8 for high lever, 0.1 for other time bins
  % M1 2: 0.7 for both lever, 0.05 for other time bins
  % M1 3: 0.5 for all the time
  M1_part = [0.1*ones(1, startTime+50+responseTime), (0.1+0.7*(trialType==2))*ones(1, pressTime);
    0.05*ones(1, startTime+50+responseTime), 0.7*ones(1, pressTime);
    0.5*ones(1, startTime+50+responseTime+pressTime)];
  
  segment_part = [zeros(1, startTime), ones(1, 50), zeros(1, responseTime), (1+trialType)*ones(1, pressTime)];
  
  trialNo_part = [zeros(1, startTime), trialIdx*ones(1, 50), zeros(1, responseTime), trialIdx*ones(1, pressTime)];
  
  rate_input(1, cursor+(1:length(mPFC_part))) = mPFC_part;
  rate_output(:, cursor+(1:length(mPFC_part))) = M1_part;
  segment(:, cursor+(1:length(mPFC_part))) = segment_part;
  trialNo(:, cursor+(1:length(mPFC_part))) = trialNo_part;
  
  cursor = cursor + length(mPFC_part);

end

rate_input = gaussianSmooth(rate_input, 10); % smooth the neural firing rate
mPFC = rate_input>rand(size(rate_input));%% Bernoulli generation
rate_output = gaussianSmooth(rate_output', 10)';  % smooth the neural firing rate
M1 = rate_output>rand(size(rate_output));%% Bernoulli generation

modelName = 'decodingModel_simulation';
his = 16;

%%
plotIndex = 1:1000;
timeIndex = plotIndex/100;
f = figure();
f.Color = 'w';
subplot(4,1,1)
plot(timeIndex, rate_input(plotIndex), 'k', 'LineWidth',1.5)
ylabel('mPFC')
box off
subplot(4,1,2)
plot(timeIndex, rate_output(1, plotIndex), 'k', 'LineWidth',1.5)
ylim([0 0.8])
box off
ylabel('M1 1')
subplot(4,1,3)
plot(timeIndex, rate_output(2, plotIndex), 'k', 'LineWidth',1.5)
ylim([0 0.8])
box off
ylabel('M1 2')
subplot(4,1,4)
plot(timeIndex, segment(plotIndex), 'k', 'LineWidth',1.5)
box off
ylabel('movements')

M1 = M1';
mPFC = mPFC';
save data\simulations.mat mPFC mPFCnum M1 M1num M1order M1num_pre segment trialNo his modelName DataIndex
